Add word count filter criterion to search

jiru opened the pull request:
Closes #1954. No reindexation required for this to work since the word count is already indexed by Manticore. Some large chunks of code changes in `tests/` are just code refactoring.

jiru committed:
Add word count filter backend

Backend code to allow filtering sentences by word count.
For languages without word boundaries, this will filter
by character count.

word_count=8                   Only sentences with 8 words.
word_count=8&word_count_op=eq  Only sentences with 8 words.
word_count=3&word_count_op=ge  Only sentences with 3 or more words.
word_count=7&word_count_op=le  Only sentences with 7 or less words.

jiru committed:
Add word count filter frontend

Add a new field in the search form that allows
setting word_count and word_count_op GET parameters.

See parent commit.

jiru committed:
Add unit tests for word count search filter

jiru committed:
Turn word count filter into a single method

Even though the code is a bit more complex, it makes
much more sense in terms of API to have a single method.
This will allow to easily add more than one word filter in
the future, such as "word_count > 6 and word_count < 12".

Also added tests for the SentencesSearchForm class.

jiru committed:
Avoid unnecessary word filter for default filters

Optimization: when no word count filter is selected,
avoid adding unnecessary filter to the Manticore query.

jiru committed:
Avoid sending string "null" as GET parameter value

Fix: when submitting a search with empty value for field word_count,
angular reads null and the parser turns that into a string "null".
Just send an empty string instead, it is cleaner.

jiru commented:
@alanfgh or anyone else: feedback welcome on wording:

![img](https://user-images.githubusercontent.com/5107734/215547284-5bf7b10c-6901-4cfd-853d-e905061e763a.png)
![img](https://user-images.githubusercontent.com/5107734/215548313-f06db0cb-77c1-4119-9332-f47988d0b7ce.png)

alanfmw commented:
The wording looks fine to me. But we could just as well start from 1 rather than 0, since we don't have any empty sentences (or shouldn't, anyway). I think that would raise fewer questions.

Thanks for asking, @jiru .

alanfgh commented:
The wording looks fine to me. But we could just as well start from 1 rather than 0, since we don't have any empty sentences (or shouldn't, anyway). I think that would raise fewer questions.

Thanks for asking, @jiru .

LBeaudoux commented:
@jiru thanks for adding this feature. 

Why not use a [range slider](https://material.angular.io/components/slider/overview#selecting-a-range)? This would be very helpful for those who prefer medium length sentences.

It would be also nice if you could set your favorite sentence length range on your settings page so you don't have to do it every time you search.

jiru commented:
@alanfgh Thanks for checking the wording. In theory, there are no 0-words sentences, but it actually sometimes happens as a result of some bug that needs to be fixed, so searching for 0-words sentences actually helps tracking such bugs. I agree it would be less confusing to default to 1 instead of 0.

@LBeaudoux About allowing both upper and lower limit, yes it would be a good improvement. About using the range slider, I am not sure of the benefits in terms of usability because the min/max numbers are not so visible. Besides, how to decide on the upper limit of the slider? By the way, Tatoeba is using [AngularJS Material](https://material.angularjs.org/), not Angular Material, so we cannot use what you linked.

About adding a new user setting, if we are to add a new setting for every little thing like that, it would become very complicated for both developers and users. I would rather try a different approach, such as improving the ranking algorithm (or adding a new one) that favor sentences that are not too long nor too short.

LBeaudoux commented:
> About allowing both upper and lower limit, yes it would be a good improvement.

What do you think of 2 custom inputs taking optional values and displaying "min." and "max." as placeholders?

Length:`min.`⊠ to `max.`⊠ word(s)

jiru committed:
Change word count filter to min/max fields

This makes the form easier to use. It is more straightforward to manipulate
and requires less interaction to accomplish word count filtering.

jiru committed:
Fix word_count_min=0 translated into 'any'

The condition for Router::url() to strip out a GET parameter is defined in
Cake\Routing\Route\Route::match(), code snippet:

            // keys that don't exist are different.
            if (!$defaultExists && ($value !== null && $value !== false && $value !== '')) {
                $query[$key] = $value;
                unset($url[$key]);
            }

So this check is different from empty().

jiru commented:
@LBeaudoux Thanks for the suggestion, it totally makes sense.

I’ve updated this pull request, the form now looks like this:
![image](https://github.com/Tatoeba/tatoeba2/assets/5107734/1d9fe612-b55d-4203-b417-75510503c955)

ckjpn commented:
Note that the word "Length:" doesn't fit.

<img width="264" alt="Screen Shot 2023-07-30 at 7 21 28" src="https://github.com/Tatoeba/tatoeba2/assets/7934273/f0be2507-b835-422a-8647-97b36ef6e0a3">

One idea that might solve this problem would  be to change "Length" to  "Length in words:" and eliminate the "word(s)" part.
This may give the same information, but properly word-wrap in the space provided.

jiru commented:
@ckjpn I updated the design on https://dev.tatoeba.org/, can you check if it still word-wraps the label on your end?

ckjpn commented:
It looks good on the several Macintosh browsers I tested it on.

ckjpn commented:
This gets unexpected results.

Length = 8, search Japanese

="何と読む"

https://dev.tatoeba.org/en/sentences/search?from=jpn&has_audio=&native=&orphans=no&query=%3D%22%E4%BD%95%E3%81%A8%E8%AA%AD%E3%82%80%22&sort=relevance&sort_reverse=&tags=&to=&trans_filter=limit&trans_has_audio=&trans_link=&trans_orphan=&trans_to=&trans_unapproved=&trans_user=&unapproved=no&user=&word_count_max=8&word_count_min=8

jiru commented:
This happens because `「` and `」` are treated as words. @ckjpn Can you open a new issue please?
